src/tensor_ops/to_dtype/mod.rs

```diff
@@ -0,0 +1,87 @@
+mod cpu_kernel;
+#[cfg(feature = "cuda")]
+mod cuda_kernel;
+
+use crate::prelude::{DeviceStorage, Shape, Tensor, Unit};
+
+pub trait ToDtypeKernel<E1: Unit, E2: Unit>: DeviceStorage {
+    fn forward<S: Shape>(inp: Tensor<S, E1, Self>) -> Result<Tensor<S, E2, Self>, Self::Err>;
+}
+
+/// Copies the elements of a tensor, converting its data to a different dtype.
+///
+/// Example usage:
+/// ```rust
+/// # use dfdx::prelude::*;
+/// # let dev: Cpu = Default::default();
+/// let a = dev.tensor([1usize, 2, 3, 4, 5]);
+///
+/// let b = a.clone().to_dtype::<i8>();
+/// assert_eq!(b.array(), [1, 2, 3, 4, 5]);
+///
+/// let c = a.clone().to_dtype::<f32>();
+/// assert_eq!(c.array(), [1.0, 2.0, 3.0, 4.0, 5.0]);
+/// ```
+pub fn to_dtype<E2: Unit, S: Shape, E1: Unit, D: ToDtypeKernel<E1, E2>>(
+    tensor: Tensor<S, E1, D>,
+) -> Tensor<S, E2, D> {
+    tensor.to_dtype()
+}
+
+impl<S: Shape, E: Unit, D: DeviceStorage> Tensor<S, E, D> {
+    pub fn try_to_dtype<E2: Unit>(self) -> Result<Tensor<S, E2, D>, D::Err>
+    where
+        D: ToDtypeKernel<E, E2>,
+    {
+        D::forward(self)
+    }
+
+    pub fn to_dtype<E2: Unit>(self) -> Tensor<S, E2, D>
+    where
+        D: ToDtypeKernel<E, E2>,
+    {
+        self.try_to_dtype().unwrap()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::{prelude::Rank1, tensor::*, tests::*};
+
+    #[test]
+    fn test_to_dtype() {
```

Comment:
So just this is increasing compile times by that much? I guess because it needs a generic version for every combo...

Is it possible to reduce the times by splitting this up into separate test functions? I'm not sure how rustc compiles test but maybe it compiles separate tests in parallel?

Reply:
That does seem to have improved compile times on my machine, but I'm not sure that it will have a benefit for github actions, as I don't think they are given access to multiple threads.

Reply:
Lucky us 😀